apps/prairielearn/src/ee/lib/validateHTML.ts

```diff
@@ -0,0 +1,602 @@
+          ];
+        }
+    }
+  }
+  return [];
+}
+
+/**
+ * Checks that a `pl-multiple-choice` element has valid properties.
+ * @param ast The tree to consider, rooted at the tag.
+ * @returns The list of errors for the tag, if any.
+ */
+function checkMultipleChoice(ast: DocumentFragment | ChildNode): string[] {
+  const errors: string[] = [];
+  let usedAnswersName = false;
+  let displayDropdown = false;
+  let usedAllOfTheAbove = false;
+  let usedNoneOfTheAbove = false;
+  let usedAllOfTheAboveFeedback = false;
+  let usedNoneOfTheAboveFeedback = false;
+  let usedSize = false;
+  const optionsOfTheAbove = ['false', 'random', 'correct', 'incorrect'];
+  if ('attrs' in ast) {
+    for (const attr of ast.attrs) {
+      const key = attr.name;
+      const val = attr.value;
+      switch (key) {
+        case 'answers-name':
+          usedAnswersName = true;
+          break;
+        case 'weight':
+          assertInt('pl-multiple-choice', key, val, errors);
+          break;
+        case 'display':
+          assertInChoices('pl-multiple-choice', key, val, ['block', 'inline', 'dropdown'], errors);
+          if (val === 'dropdown') {
+            displayDropdown = true;
+          }
+          break;
+        case 'number-answers':
+          assertInt('pl-multiple-choice', key, val, errors);
+          break;
+        case 'order':
+          assertInChoices(
+            'pl-multiple-choice',
+            key,
+            val,
+            ['random', 'ascend', 'descend', 'fixed'],
+            errors,
+          );
+          break;
+        case 'hide-letter-keys':
+          assertBool('pl-multiple-choice', key, val, errors);
+          break;
+        case 'all-of-the-above':
+          assertInChoices('pl-multiple-choice', key, val, optionsOfTheAbove, errors);
+          if (optionsOfTheAbove.includes(val) && val !== 'false') {
+            usedAllOfTheAbove = true;
+          }
+          break;
+        case 'none-of-the-above':
+          assertInChoices('pl-multiple-choice', key, val, optionsOfTheAbove, errors);
+          if (optionsOfTheAbove.includes(val) && val !== 'false') {
+            usedNoneOfTheAbove = true;
+          }
+          break;
+        case 'all-of-the-above-feedback':
+          usedAllOfTheAboveFeedback = true;
+          break;
+        case 'none-of-the-above-feedback':
+          usedNoneOfTheAboveFeedback = true;
+          break;
+        case 'allow-blank':
+          assertBool('pl-multiple-choice', key, val, errors);
+          break;
+        case 'size':
+          usedSize = true;
+          assertInt('pl-multiple-choice', key, val, errors);
+          break;
+        default:
+          if (!htmlGlobalAttributes.includes(key)) {
+            errors.push(`pl-multiple-choice: ${key} is not a valid property.`);
+          }
```

Comment:
When are any of these attributes ever valid for a PL element?

Reply:
The LLM was generating `id`s and `class`es for tags; I think that's legal but I can remove it if not.

Reply:
That will fail at runtime, I'd encourage you to test that for yourself. Consider the following example:

```html
<pl-string-input answers-name="plural_word" label="Plural" remove-spaces="true" class="not-valid"></pl-string-input>
```

Generating a question with that yields this error:

```python
Traceback (most recent call last):
  File "/Users/nathan/git/PrairieLearn/apps/prairielearn/python/zygote.py", line 416, in <module>
    worker_loop()
  File "/Users/nathan/git/PrairieLearn/apps/prairielearn/python/zygote.py", line 324, in worker_loop
    val = method(*args)
  File "/Users/nathan/git/PrairieLearn/apps/prairielearn/elements/pl-string-input/pl-string-input.py", line 53, in prepare
    pl.check_attribs(element, required_attribs, optional_attribs)
  File "/Users/nathan/git/PrairieLearn/apps/prairielearn/python/prairielearn.py", line 506, in check_attribs
    raise Exception('Unknown attribute "%s"' % name)
Exception: Unknown attribute "class"
Traceback (most recent call last):
  File "/Users/nathan/git/PrairieLearn/apps/prairielearn/python/zygote.py", line 458, in <module>
    raise Exception(
Exception: worker process exited unexpectedly with status 256
```

`check_attribs` is strict and will reject anything that isn't specifically allowed.

Reply:
changed in 1c94d0e0b1e10d529169226b67b29884d6c7cd49